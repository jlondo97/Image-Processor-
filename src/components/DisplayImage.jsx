/* eslint-disable jsx-a11y/label-has-associated-control */
import React from 'react';
import PropTypes from 'prop-types';

function DisplayImage({
  image,
  width,
  height,
  containerWidth,
  containerHeight,

}) {
  return (
    <div>
      <div
        className="card mb-3"
        id="HojaCarta"
        style={{
          margin: 0,
          padding: 0,
          width: containerWidth,
          height: containerHeight,
        }}
      >

        <img src={image} style={{ width, height }} className="rounded" alt="" />
      </div>
    </div>

  );
}

DisplayImage.propTypes = {
  image: PropTypes.string,
  width: PropTypes.number,
  height: PropTypes.number,
  containerHeight: PropTypes.number,
  containerWidth: PropTypes.number,

};

DisplayImage.defaultProps = {
  image: null,
  width: 0,
  height: 0,
  containerWidth: 796,
  containerHeight: 1123,
};
export default DisplayImage;
