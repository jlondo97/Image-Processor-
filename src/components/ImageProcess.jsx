/* eslint-disable jsx-a11y/label-has-associated-control */
import React from 'react';
import DisplayImage from './DisplayImage';

const DECIMAL_BASE = 10;
const INITIAL_WIDTH = 796;
const INITIAL_HEIGHT = 1123;

class ImageProcess extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      height: null,
      image: null,
      isToggleOn: true,
      width: null,
      imageprocess: null,
      process: false,
      containerWidth: INITIAL_WIDTH,
      containerHeigth: INITIAL_HEIGHT,
    };

    this.handleClick = this.handleClick.bind(this);
    this.onImageChange = this.onImageChange.bind(this);
  }

  handleClick() {
    this.setState((state) => ({
      isToggleOn: !state.isToggleOn,
    }));
    this.convertSize();
  }

  onImageChange(event) {
    if (event.target.files && event.target.files[0]) {
      const img = new Image();
      const src = URL.createObjectURL(event.target.files[0]);

      img.onload = () => {
        const { height, width } = img;
        this.setState({
          height,
          width,
          process: false,
        });
      };

      img.src = src;

      this.setState({
        image: src,
      });
    }
  }

  get orientation() {
    let orientation = 'Cuadrado';
    if (this.state.width > this.state.height) {
      orientation = 'Horizontal';
    }
    if (this.state.width < this.state.height) {
      orientation = 'Vertical';
    }
    return orientation;
  }

  get imagesize() {
    const { height, width } = this.state;
    return height && width && ` ${parseInt(width, DECIMAL_BASE)} x ${parseInt(height, DECIMAL_BASE)} , Orientacion ${this.orientation} , Proporcion ${this.proportion}`;
  }

  get proportion() {
    const { height, width } = this.state;
    const gcd = this.aspectRatio(width, height);
    return (`${width / gcd} : ${height / gcd}`);
  }

  get imageprocess() {
    const { process, imageprocess, image } = this.state;
    return process ? imageprocess : image;
  }

  aspectRatio(a, b) {
    return (b === 0) ? a : this.aspectRatio(b, a % b);
  }

  convertSize() {
    const { image, containerHeigth, containerWidth } = this.state;
    const imageprocess = new Image();
    let { width, height } = this.state;
    let maxHeight = containerHeigth;
    let maxWidth = containerWidth;
    let ratio = 0;

    imageprocess.src = image;

    if (this.orientation === 'Horizontal') {
      maxWidth = containerHeigth;
      maxHeight = containerWidth;
      this.setState({
        containerHeigth: INITIAL_WIDTH,
        containerWidth: INITIAL_HEIGHT,
      });
    }

    if (this.orientation === 'Vertical') {
      maxWidth = INITIAL_WIDTH;
      maxHeight = INITIAL_HEIGHT;
      this.setState({
        containerHeigth: INITIAL_HEIGHT,
        containerWidth: INITIAL_WIDTH,
      });
    }
    // eslint-disable-next-line no-debugger
    // debugger;
    if (width > maxWidth) {
      ratio = maxWidth / width;
      height *= ratio; // Reset height to match scaled image
      width *= ratio; // Reset width to match scaled image
    }

    // Check if current height is larger than max
    if (height > maxHeight) {
      ratio = maxHeight / height;
      width *= ratio; // Reset width to match scaled image
      height *= ratio; // Reset height to match scaled image
    }

    this.setState({
      width,
      height,
      process: true,
      isToggleOn: true,
      // containerHeigth: maxHeight,
      // containerWidth: maxWidth,
      imageprocess: imageprocess.src,
    });
  }

  render() {
    return (
      <div className="container text-white">
        <div className="input-group">
          <div className="custom-file">
            <input type="file" className="custom-file-input" id="inputGroupFile04" aria-describedby="inputGroupFileAddon04" onChange={this.onImageChange} />
            <label className="custom-file-label" htmlFor="inputGroupFile04">Sube tu imagen</label>
          </div>
          <div className=" outline-secondary">
            <button className="btn btn-light" onClick={this.handleClick} type="button" id="inputGroupFileAddon04">
              {this.state.isToggleOn ? 'Procesar' : 'Procesando'}
            </button>
          </div>
        </div>
        <div className="card-body" id="div1">
          {this.imagesize}
          <DisplayImage
            image={this.imageprocess}
            width={this.state.width}
            height={this.state.height}
            containerHeight={this.state.containerHeigth}
            containerWidth={this.state.containerWidth}
          />

        </div>

      </div>
    );
  }
}

export default ImageProcess;
