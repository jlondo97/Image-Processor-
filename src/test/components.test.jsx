import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import DisplayImage from '../components/DisplayImage';
import ImageProcess from '../components/ImageProcess';

configure({ adapter: new Adapter() });

describe('DisplayImage SuitTest', () => {
  test('render DisplayImage', () => {
    const component = renderer.create(
      <DisplayImage
        image="https://cumbrepuebloscop20.org/wp-content/uploads/2018/09/Oso-Pardo-1.jpg"
        width={1140}
        height={654}
      />,
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('ImageProcess SuitTest', () => {
  test('convertSize is sucessfully ', () => {
    const component = shallow(
      <ImageProcess />,
    );
    const initialState = {
      image: 'https://cumbrepuebloscop20.org/wp-content/uploads/2018/09/Oso-Pardo-1.jpg',
      isToggleOn: true,
      width: 1140,
      height: 654,
      imageprocess: null,
      process: false,
    };
    component.setState(initialState);
    const expectWidth = 1123;
    const expectHeigth = 644;
    component.instance().convertSize();

    expect(parseInt(component.state('width'), 10)).toEqual(expectWidth);
    expect(parseInt(component.state('height'), 10)).toEqual(expectHeigth);
  });
});
