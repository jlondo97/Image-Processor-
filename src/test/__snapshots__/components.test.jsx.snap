// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DisplayImage SuitTest render DisplayImage 1`] = `
<div>
  <div
    className="card mb-3"
    id="HojaCarta"
    style={
      Object {
        "height": 1123,
        "margin": 0,
        "padding": 0,
        "width": 796,
      }
    }
  >
    <img
      alt=""
      className="rounded"
      src="https://cumbrepuebloscop20.org/wp-content/uploads/2018/09/Oso-Pardo-1.jpg"
      style={
        Object {
          "height": 654,
          "width": 1140,
        }
      }
    />
  </div>
</div>
`;
